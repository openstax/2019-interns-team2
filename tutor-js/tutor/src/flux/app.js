// Store for api requests and responses
import _ from 'underscore';

import { makeSimpleStore } from './helpers';


const AppConfig = {
  resetServerErrors() {
    return delete this._currentServerError;
  },

  setServerError(errorResponse) {
    const { status, data } = errorResponse;
    this._currentServerError = errorResponse;
    return this.emit('server-error', status, data);
  },

  exports: {
    getError() { return this._currentServerError; },

    errorNavigation() {
      if (!this._currentServerError) { return {}; }
      const { status, request } = this._currentServerError;
      if (status === 403) {
        return { href: '/' };
      } else {
        const isGET404 = (status === 404) && (request.method === 'GET');
        const isInRange = 400 <= status && status < 600;
        return { shouldReload: false };
      }
    },
  },
};

const { actions, store } = makeSimpleStore(AppConfig);
export { actions as AppActions, store as AppStore };
