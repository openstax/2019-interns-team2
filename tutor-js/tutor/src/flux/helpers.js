import _ from 'underscore';
import flux from 'flux-react';

const LOADING = 'loading';
const LOADED  = 'loaded';
const FAILED  = 'failed';
const SAVING  = 'saving';
const DELETING = 'deleting';
const DELETED = 'deleted';

const STATES = {
  LOADING,
  LOADED,
  FAILED,
  SAVING,
  DELETING,
  DELETED,
};

let idCounter = 0;
const CREATE_KEY = () => `_CREATING_${idCounter++}`;

const isNew = id => /_CREATING_/.test(id);

const CrudConfig = function() {
  return {
    _asyncStatus: {},
    _local: {},
    _changed: {},
    _errors: {},
    _reload: {},

    // If the specific type needs to do something else to the object:
    // _loaded : (obj, id) ->
    // _saved : (obj, id) ->
    // _reset : ->

    reset() {
      this._asyncStatus = {};
      this._local = {};
      this._changed = {};
      this._errors = {};
      this._reload = {};

      if (typeof this._reset === 'function') {
        this._reset();
      }
      return this.emitChange();
    },

    dontReload(id) {
      return (this._asyncStatus[id] === LOADED) && this._HACK_DO_NOT_RELOAD;
    },

    FAILED(status, msg, id, ...args) {
      this._asyncStatus[id] = FAILED;
      this._errors[id] = msg;
      if (typeof this._failed === 'function') {
        this._failed({ status, msg }, id, ...Array.from(args));
      }

      if (status !== 0) { // indicates network failure
        delete this._local[id];
        return this.emitChange();
      }
    },

    load(id) {
      return this._load(id);
    },

    // allow aliasing load without triggering api call.
    _load(id) {
      // Add a shortcut for unit testing
      if (this.dontReload(id)) { return; }
      this._reload[id] = false;
      this._asyncStatus[id] = LOADING;
      return this.emitChange();
    },

    loaded(obj, id, ...args) {
      // id = obj.id
      this._asyncStatus[id] = LOADED;

      if (obj) {
        // If the specific type needs to do something else to the object:
        this._local[id] = (typeof this._loaded === 'function' ? this._loaded(obj, id, ...Array.from(args)) : undefined) || obj;
      }

      return this.emitChange();
    },

    save(id, obj) {
      // Note: id could be isNew()
      this._asyncStatus[id] = SAVING;
      return this.emitChange();
    },

    saved(result, id, ...args) {
      // id = result.id
      this._asyncStatus[id] = LOADED; // TODO: Maybe make this SAVED

      // If the specific type needs to do something else to the object:
      const obj = typeof this._saved === 'function' ? this._saved(result, id, ...Array.from(args)) : undefined;
      if (obj) { result = obj; }

      if (result) {
        this._local[id] = result;
        this._local[result.id] = result;
        delete this._changed[result.id];
      } else {
        console.warn('API WARN: Server did not return JSON after saving. Patching locally');
        // Merge all the local changes into the new local object
        this._local[id] = _.extend(this._local[id], this._changed[id]);
      }
      delete this._changed[id];
      delete this._errors[id];
      // If the specific type needs to do something else to the object:
      return this.emitChange();
    },

    create(localId, attributes = {}) {
      if (!isNew(localId)) { throw new Error('BUG: MUST provide a local id'); }
      this._local[localId] = {};
      this._changed[localId] = attributes;
      return this._asyncStatus[localId] = LOADED;
    },

    created(result, localId, ...args) {
      this._local[localId] = result; // HACK: So react component can still manipulate the same object
      this._local[result.id] = result;
      this._asyncStatus[localId] = LOADED;
      this._asyncStatus[result.id] = LOADED;
      this.emitChange();
      return (typeof this._created === 'function' ? this._created(result, localId, ...Array.from(args)) : undefined);
    },

    _change(id, obj) {
      if (this._changed[id] == null) { this._changed[id] = {}; }
      _.extend(this._changed[id], obj);
      return this.emitChange();
    },

    _save(id) {
      return this._asyncStatus[id] = SAVING;
    },

    delete(id) {
      this._asyncStatus[id] = DELETING;
      return (typeof this._delete === 'function' ? this._delete(id) : undefined);
    },

    deleted(result, id, ...args) {
      this._asyncStatus[id] = DELETED;
      delete this._local[id];
      this.emitChange();
      return (typeof this._deleted === 'function' ? this._deleted(result, id, ...Array.from(args)) : undefined);
    },

    clearChanged(id) {
      return delete this._changed[id];
    },

    HACK_DO_NOT_RELOAD(bool) { return this._HACK_DO_NOT_RELOAD = bool; },

    // Keep this here so other exports method have access to it
    _get(id) {
      const val = this._local[id];
      if (!val && (this._asyncStatus[id] !== SAVING)) { return null; }
      // Scores stores an Array unlike most other stores
      if (_.isArray(val)) {
        return val;
      } else {
        return _.extend({}, val, this._changed[id]);
      }
    },

    exports: {
      isUnknown(id) { return !this._asyncStatus[id]; },
      isLoading(id) { return this._asyncStatus[id] === LOADING; },
      isLoaded(id) { return this._asyncStatus[id] === LOADED; },
      isDeleting(id) { return this._asyncStatus[id] === DELETING; },
      isSaving(id) { return this._asyncStatus[id] === SAVING; },
      isFailed(id) { return this._asyncStatus[id] === FAILED; },
      getAsyncStatus(id) { return this._asyncStatus[id]; },
      get(id) { return this._get(id); },
      isChanged(id) { return !_.isEmpty(this._changed[id]); },
      getChanged(id) { return this._changed[id] || {}; },
      freshLocalId() { return CREATE_KEY(); },
      isNew(id) { return isNew(id); },
      reload(id) { return this._reload[id]; },
      getError(id) { return this._errors[id]; },
    },
  };
};

// Helper for creating a simple store for actions
const makeSimpleStore = function(storeConfig) {

  let actionsConfig = _.omit(storeConfig, (value, key) => (typeof value !== 'function') || (key === 'exports'));

  actionsConfig = _.keys(actionsConfig);
  const actions = flux.createActions(actionsConfig);
  storeConfig.actions = _.values(actions);
  const store = flux.createStore(storeConfig);
  return { actions, store };
};


const extendConfig = function(newConfig, origConfig) {
  if (newConfig.exports == null) { newConfig.exports = {}; }
  _.defaults(newConfig, origConfig);
  _.defaults(newConfig.exports, origConfig.exports);
  return newConfig;
};


const makeStandardStore = function(name, storeConfig) {
  extendConfig(storeConfig, new CrudConfig());
  const { actions, store } = makeSimpleStore(storeConfig);
  return { [`${name}Actions`]: actions, [`${name}Store`]: store };
};


export { CrudConfig, makeSimpleStore, extendConfig, makeStandardStore, isNew, STATES };
