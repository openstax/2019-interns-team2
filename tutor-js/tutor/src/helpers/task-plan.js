import { TaskPlanStore } from '../flux/task-plan';
import { isEmpty, get, pick, first, every, reduce, map } from 'lodash';

export default {

  // returns the opens_at and due_at dates combined if they're identical,
  // separated by target_id if not
  dates(plan, options) {
    if (options == null) { options = {}; }
    if (isEmpty( get(plan, 'tasking_plans' ))) { return {}; }
    const pickFn = options.only ? tp => pick(tp, options.only) : tp => pick(tp, 'opens_at', 'due_at');

    const firstTasking = first( plan.tasking_plans );
    const isIdentical = every(plan.tasking_plans, function(tp) {
      if (options.only != null) {
        return tp[options.only] === firstTasking[options.only];
      } else {
        return (tp.opens_at === firstTasking.opens_at) && (tp.due_at === firstTasking.due_at);
      }
    });
    if (isIdentical) {
      return { all: pickFn(firstTasking) };
    } else {
      return reduce( plan.tasking_plans, function(memo, tp) {
        memo[tp.target_id] = pickFn(tp);
        return memo;
      }
      , {});
    }
  },


  earliestDueDate(plan) {
    const dates = map(get(plan, 'tasking_plans'), 'due_at');
    return first(dates.sort()) || '';
  },


  apiEndpointOptions(id, courseId) {
    const task = TaskPlanStore.get(id);
    const options = TaskPlanStore.isNew(id) ? {
      url: `courses/${courseId}/plans`,
      method: 'POST',
    } : {
      url: `plans/${id}`,
    };
    if (get(task, 'ecosystem_id')) {
      options.params = { ecosystem_id: task.ecosystem_id };
    }
    return options;
  },

};
