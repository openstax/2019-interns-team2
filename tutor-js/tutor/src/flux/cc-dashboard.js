import _ from 'underscore';
import { CrudConfig, makeSimpleStore, extendConfig } from './helpers';
import PeriodHelper from '../helpers/period';

const DEFAULT_COURSE_TIMEZONE = 'US/Central';

const DashboardConfig = {
  exports: {
    isBlank(courseId) {
      return !_.any(__guard__(__guard__(this._get(courseId), x1 => x1.course), x => x.periods));
    },

    getPeriods(courseId) { return PeriodHelper.sort(__guard__(__guard__(this._get(courseId), x1 => x1.course), x => x.periods)); },

    chaptersForDisplay(courseId, periodId) {
      const period = _.findWhere( __guard__(__guard__(this._get(courseId), x1 => x1.course), x => x.periods), { id: periodId });
      if (!period) { return []; }
      for (let chapter of period.chapters) {
        for (let page of chapter.pages) {
          const total = page.completed + page.in_progress + page.not_started;
          page.completed_percentage = page.completed / total;
        }
        chapter.valid_sections = _.sortBy(chapter.pages, page => page.chapter_section[1] || 0).reverse();
        chapter;
      }

      return _.select(period.chapters.reverse(), chapter => chapter.valid_sections != null ? chapter.valid_sections.length : undefined);
    },
  },
};


extendConfig(DashboardConfig, new CrudConfig());
const { actions, store } = makeSimpleStore(DashboardConfig);
export { actions as CCDashboardActions, store as CCDashboardStore };

function __guard__(value, transform) {
  return (typeof value !== 'undefined' && value !== null) ? transform(value) : undefined;
}
