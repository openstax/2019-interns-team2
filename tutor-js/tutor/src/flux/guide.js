import { CrudConfig, makeSimpleStore, extendConfig } from './helpers';

const CourseGuideConfig = {};

extendConfig(CourseGuideConfig, new CrudConfig());
const { actions, store } = makeSimpleStore(CourseGuideConfig);
export { actions as CourseGuideActions, store as CourseGuideStore };
