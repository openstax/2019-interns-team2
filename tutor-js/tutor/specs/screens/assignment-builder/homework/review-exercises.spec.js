import { React, C } from '../../../helpers';
import { map } from 'lodash';
import ReviewExercises from '../../../../src/screens/assignment-builder/homework/review-exercises';
import Factory, { FactoryBot } from '../../../factories';
import FakeWindow from 'shared/specs/helpers/fake-window';
import { ExtendBasePlan } from '../task-plan-helper';
import { TaskPlanStore } from '../../../../src/flux/task-plan';

const PLAN_ID  = '1';
const NEW_PLAN = ExtendBasePlan({ id: PLAN_ID });

jest.mock('react-dom');
jest.mock('../../../../src/flux/task-plan');
jest.mock('../../../../../shared/src/components/html', () => ({ html }) =>
  html ? <div dangerouslySetInnerHTML={{ __html: html }} /> : null
);

describe('choose exercises component', function() {
  let exercises, props, course;

  beforeEach(function() {
    course = Factory.course();
    course.referenceBook.onApiRequestComplete({ data: [FactoryBot.create('Book')] });
    exercises = Factory.exercisesMap({ book: course.referenceBook });
    const pageIds = course.referenceBook.children[1].children.map(pg => pg.id);
    const items = pageIds.map(page_id =>
      FactoryBot.create('TutorExercise', { page_uuid: course.referenceBook.pages.byId.get(page_id).uuid }),
    );
    exercises.onLoaded({ data: { items } }, [{ course, page_ids: pageIds }]);
    TaskPlanStore.getExercises.mockImplementation(() => map(exercises.array, 'id'));

    props = {
      course,
      exercises,
      pageIds,
      onAddClick: jest.fn(),
      windowImpl: new FakeWindow,
      showSectionTopics: jest.fn(),
      canEdit: false,
      planId: PLAN_ID,
      cancel: jest.fn(),
      hide: jest.fn(),
    };
  });

  it('matches snapshot', () => {
    expect.snapshot(<C><ReviewExercises {...props} /></C>).toMatchSnapshot();
  });

});
