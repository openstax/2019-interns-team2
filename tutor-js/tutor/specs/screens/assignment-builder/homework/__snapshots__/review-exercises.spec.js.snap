// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`choose exercises component matches snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
}

<div
  className="openstax-debug-content"
>
  <div
    data-purpose="tour-conductor-wrapper"
  >
    <div>
      <div
        className="homework-builder-view"
      >
        <div
          className="exercise-controls-bar"
        >
          <div
            className="controls"
          />
          <div
            className="indicators"
          >
            <div
              className="num total"
            >
              <h2>
                NaN
              </h2>
              <span>
                Total Problems
              </span>
            </div>
            <div
              className="num mine"
            >
              <h2>
                0
              </h2>
              <span>
                My Selections
              </span>
            </div>
            <div
              className="tour-anchor num tutor"
              data-tour-anchor-id="tutor-selections"
            >
              <div
                className="tutor-selections"
              >
                <span
                  className="circle-btn-placeholder"
                />
                <h2 />
                <span
                  className="circle-btn-placeholder"
                />
              </div>
              <span>
                OpenStax Tutor Selections
              </span>
              <button
                className="btn btn-link"
                disabled={false}
                id="homework-selections-trigger"
                onClick={[Function]}
                type="button"
              >
                What are these?
              </button>
            </div>
            <div
              className="tutor-added-later"
            >
              <span>
                Tutor selections are added later to support spaced practice and personalized learning.
              </span>
            </div>
          </div>
          <div
            className="actions"
          />
        </div>
        <table
          className="exercise-table"
          data-tour-region-id="add-homework-review-sections"
        >
          <thead>
            <tr>
              <td />
              <td />
              <td>
                Problem Question
              </td>
              <td>
                Learning Objective
              </td>
              <td>
                Details
              </td>
            </tr>
          </thead>
          <tbody>
            <tr
              data-ex-id={1}
            >
              <td
                className="exercise-number"
              >
                1
              </td>
              <td>
                <span
                  className="chapter-section"
                  data-chapter-section=".0"
                >
                  
                </span>
              </td>
              <td>
                <div
                  dangerouslySetInnerHTML={
                    Object {
                      "__html": "What conditions imply that we can use classical physics without considering special relativity or quantum mechanics?",
                    }
                  }
                />
              </td>
              <td
                className="ellipses"
              >
                lo:stax-phys:1-2-1
              </td>
              <td
                className="ellipses"
              >
                blooms:3 / dok:3
              </td>
            </tr>
            <tr
              data-ex-id={2}
            >
              <td
                className="exercise-number"
              >
                2
              </td>
              <td>
                <span
                  className="chapter-section"
                  data-chapter-section=".0"
                >
                  
                </span>
              </td>
              <td>
                <div
                  dangerouslySetInnerHTML={
                    Object {
                      "__html": "How could physics be useful in weather prediction?",
                    }
                  }
                />
              </td>
              <td
                className="ellipses"
              >
                lo:stax-phys:1-2-1
              </td>
              <td
                className="ellipses"
              >
                blooms:3 / dok:3
              </td>
            </tr>
            <tr
              data-ex-id={3}
            >
              <td
                className="exercise-number"
              >
                3
              </td>
              <td>
                <span
                  className="chapter-section"
                  data-chapter-section=".0"
                >
                  
                </span>
              </td>
              <td>
                <div
                  dangerouslySetInnerHTML={
                    Object {
                      "__html": "In what sense does Einsteinâ€™s theory of relativity illustrate that physics describes fundamental aspects of our universe?",
                    }
                  }
                />
              </td>
              <td
                className="ellipses"
              >
                lo:stax-phys:1-2-1
              </td>
              <td
                className="ellipses"
              >
                blooms:3 / dok:3
              </td>
            </tr>
          </tbody>
        </table>
        <div
          className="card-list exercises"
        >
          <div
            className="openstax exercise-wrapper"
          >
            <div
              className="openstax-exercise-preview exercise-card non-interactive is-vertically-truncated card"
              data-exercise-id="1@1"
              tabIndex={-1}
            >
              <div
                className="card-header"
              >
                <div
                  className="-exercise-header c0"
                >
                  <span
                    className="exercise-number"
                  >
                    1
                  </span>
                </div>
              </div>
              <div
                className="card-body"
              >
                <div
                  className="exercise-body"
                >
                  <div
                    className="openstax-question openstax-question-preview"
                  >
                    <div
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "What conditions imply that we can use classical physics without considering special relativity or quantum mechanics?",
                        }
                      }
                    />
                    <div
                      className="answers-table"
                    >
                      <div
                        className="openstax-answer"
                      >
                        <section
                          className="answers-answer disabled"
                          role="region"
                        >
                          <div
                            className="correct-incorrect"
                          />
                          <label
                            className="answer-label"
                            htmlFor="1-option-0"
                            onKeyPress={[Function]}
                          >
                            <button
                              aria-label="Choice a:"
                              className="answer-letter"
                              disabled={true}
                            >
                              a
                            </button>
                            <div
                              className="answer-answer"
                            >
                              <div
                                dangerouslySetInnerHTML={
                                  Object {
                                    "__html": "<ol>
  <li>matter is moving at speeds of less than roughly 1% the speed of light,</li>
  <li>objects are too small to be seen with the naked eye, and</li>
  <li>there is the involvement of only a weak gravitational field</li>
</ol>
",
                                  }
                                }
                              />
                            </div>
                          </label>
                        </section>
                      </div>
                      <div
                        className="openstax-answer"
                      >
                        <section
                          className="answers-answer disabled"
                          role="region"
                        >
                          <div
                            className="correct-incorrect"
                          />
                          <label
                            className="answer-label"
                            htmlFor="1-option-1"
                            onKeyPress={[Function]}
                          >
                            <button
                              aria-label="Choice b:"
                              className="answer-letter"
                              disabled={true}
                            >
                              b
                            </button>
                            <div
                              className="answer-answer"
                            >
                              <div
                                dangerouslySetInnerHTML={
                                  Object {
                                    "__html": "<ol>
  <li>matter is moving at speeds greater than roughly 1% the speed of light,</li>
  <li>objects are large enough to be seen with the naked eye, and</li>
  <li>there is the involvement of a strong gravitational field</li>
</ol>
",
                                  }
                                }
                              />
                            </div>
                          </label>
                        </section>
                      </div>
                      <div
                        className="openstax-answer"
                      >
                        <section
                          className="answers-answer disabled"
                          role="region"
                        >
                          <div
                            className="correct-incorrect"
                          />
                          <label
                            className="answer-label"
                            htmlFor="1-option-2"
                            onKeyPress={[Function]}
                          >
                            <button
                              aria-label="Choice c:"
                              className="answer-letter"
                              disabled={true}
                            >
                              c
                            </button>
                            <div
                              className="answer-answer"
                            >
                              <div
                                dangerouslySetInnerHTML={
                                  Object {
                                    "__html": "<ol>
  <li>matter is moving at speeds of less than roughly 1% the speed of light,</li>
  <li>objects are large enough to be seen with the naked eye, and</li>
  <li>there is the involvement of a strong gravitational field</li>
</ol>
",
                                  }
                                }
                              />
                            </div>
                          </label>
                        </section>
                      </div>
                      <div
                        className="openstax-answer"
                      >
                        <section
                          className="answers-answer disabled"
                          role="region"
                        >
                          <div
                            className="correct-incorrect"
                          />
                          <label
                            className="answer-label"
                            htmlFor="1-option-3"
                            onKeyPress={[Function]}
                          >
                            <button
                              aria-label="Choice d:"
                              className="answer-letter"
                              disabled={true}
                            >
                              d
                            </button>
                            <div
                              className="answer-answer"
                            >
                              <div
                                dangerouslySetInnerHTML={
                                  Object {
                                    "__html": "<ol>
  <li>matter is moving at speeds of less than roughly 1% the speed of light,</li>
  <li>objects are large enough to be seen with the naked eye, and</li>
  <li>there is the involvement of a weak gravitational field</li>
</ol>
",
                                  }
                                }
                              />
                            </div>
                          </label>
                        </section>
                      </div>
                      <div
                        className="openstax-answer"
                      >
                        <section
                          className="answers-answer disabled"
                          role="region"
                        >
                          <div
                            className="correct-incorrect"
                          />
                          <label
                            className="answer-label"
                            htmlFor="1-option-4"
                            onKeyPress={[Function]}
                          >
                            <button
                              aria-label="Choice e:"
                              className="answer-letter"
                              disabled={true}
                            >
                              e
                            </button>
                            <div
                              className="answer-answer"
                            >
                              <div
                                dangerouslySetInnerHTML={
                                  Object {
                                    "__html": "<ol>
  <li>matter is moving at speeds of less than roughly 1% the speed of light,</li>
  <li>objects are too small to be seen with the naked eye, and</li>
  <li>there is the involvement of only a weak gravitational field</li>
</ol>
",
                                  }
                                }
                              />
                            </div>
                          </label>
                        </section>
                      </div>
                    </div>
                    <div
                      className="detailed-solution"
                    >
                      <div
                        className="header"
                      >
                        Detailed solution
                      </div>
                      <div
                        dangerouslySetInnerHTML={
                          Object {
                            "__html": "four",
                          }
                        }
                      />
                    </div>
                  </div>
                </div>
                <div
                  className="exercise-tags"
                >
                  <span
                    className="exercise-tag"
                  >
                    lo:stax-phys:1-2-1
                  </span>
                  <span
                    className="exercise-tag"
                  >
                    blooms:3
                  </span>
                  <span
                    className="exercise-tag"
                  >
                    dok:3
                  </span>
                  <span
                    className="exercise-tag"
                  >
                    time:long
                  </span>
                  <span
                    className="exercise-tag"
                  >
                    ID:1@1
                  </span>
                </div>
              </div>
            </div>
          </div>
          <div
            className="openstax exercise-wrapper"
          >
            <div
              className="openstax-exercise-preview exercise-card non-interactive is-vertically-truncated card"
              data-exercise-id="2@1"
              tabIndex={-1}
            >
              <div
                className="card-header"
              >
                <div
                  className="-exercise-header c0"
                >
                  <span
                    className="exercise-number"
                  >
                    2
                  </span>
                </div>
              </div>
              <div
                className="card-body"
              >
                <div
                  className="exercise-body"
                >
                  <div
                    className="openstax-question openstax-question-preview"
                  >
                    <div
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "How could physics be useful in weather prediction?",
                        }
                      }
                    />
                    <div
                      className="answers-table"
                    >
                      <div
                        className="openstax-answer"
                      >
                        <section
                          className="answers-answer disabled"
                          role="region"
                        >
                          <div
                            className="correct-incorrect"
                          />
                          <label
                            className="answer-label"
                            htmlFor="2-option-0"
                            onKeyPress={[Function]}
                          >
                            <button
                              aria-label="Choice a:"
                              className="answer-letter"
                              disabled={true}
                            >
                              a
                            </button>
                            <div
                              className="answer-answer"
                            >
                              <div
                                dangerouslySetInnerHTML={
                                  Object {
                                    "__html": "Physics helps in predicting dynamics and movement of weather phenomena.",
                                  }
                                }
                              />
                            </div>
                          </label>
                        </section>
                      </div>
                      <div
                        className="openstax-answer"
                      >
                        <section
                          className="answers-answer disabled"
                          role="region"
                        >
                          <div
                            className="correct-incorrect"
                          />
                          <label
                            className="answer-label"
                            htmlFor="2-option-1"
                            onKeyPress={[Function]}
                          >
                            <button
                              aria-label="Choice b:"
                              className="answer-letter"
                              disabled={true}
                            >
                              b
                            </button>
                            <div
                              className="answer-answer"
                            >
                              <div
                                dangerouslySetInnerHTML={
                                  Object {
                                    "__html": "Physics helps in predicting how burning fossil fuel releases pollutants.",
                                  }
                                }
                              />
                            </div>
                          </label>
                        </section>
                      </div>
                      <div
                        className="openstax-answer"
                      >
                        <section
                          className="answers-answer disabled"
                          role="region"
                        >
                          <div
                            className="correct-incorrect"
                          />
                          <label
                            className="answer-label"
                            htmlFor="2-option-2"
                            onKeyPress={[Function]}
                          >
                            <button
                              aria-label="Choice c:"
                              className="answer-letter"
                              disabled={true}
                            >
                              c
                            </button>
                            <div
                              className="answer-answer"
                            >
                              <div
                                dangerouslySetInnerHTML={
                                  Object {
                                    "__html": "Physics helps in predicting how the flowing water affects Earthâ€™s surface.",
                                  }
                                }
                              />
                            </div>
                          </label>
                        </section>
                      </div>
                      <div
                        className="openstax-answer"
                      >
                        <section
                          className="answers-answer disabled"
                          role="region"
                        >
                          <div
                            className="correct-incorrect"
                          />
                          <label
                            className="answer-label"
                            htmlFor="2-option-3"
                            onKeyPress={[Function]}
                          >
                            <button
                              aria-label="Choice d:"
                              className="answer-letter"
                              disabled={true}
                            >
                              d
                            </button>
                            <div
                              className="answer-answer"
                            >
                              <div
                                dangerouslySetInnerHTML={
                                  Object {
                                    "__html": "Physics helps in predicting the motion of tectonic plates.",
                                  }
                                }
                              />
                            </div>
                          </label>
                        </section>
                      </div>
                      <div
                        className="openstax-answer"
                      >
                        <section
                          className="answers-answer disabled"
                          role="region"
                        >
                          <div
                            className="correct-incorrect"
                          />
                          <label
                            className="answer-label"
                            htmlFor="2-option-4"
                            onKeyPress={[Function]}
                          >
                            <button
                              aria-label="Choice e:"
                              className="answer-letter"
                              disabled={true}
                            >
                              e
                            </button>
                            <div
                              className="answer-answer"
                            >
                              <div
                                dangerouslySetInnerHTML={
                                  Object {
                                    "__html": "Physics helps in predicting dynamics and movement of weather phenomena.",
                                  }
                                }
                              />
                            </div>
                          </label>
                        </section>
                      </div>
                    </div>
                    <div
                      className="detailed-solution"
                    >
                      <div
                        className="header"
                      >
                        Detailed solution
                      </div>
                      <div
                        dangerouslySetInnerHTML={
                          Object {
                            "__html": "four",
                          }
                        }
                      />
                    </div>
                  </div>
                </div>
                <div
                  className="exercise-tags"
                >
                  <span
                    className="exercise-tag"
                  >
                    lo:stax-phys:1-2-1
                  </span>
                  <span
                    className="exercise-tag"
                  >
                    blooms:3
                  </span>
                  <span
                    className="exercise-tag"
                  >
                    dok:3
                  </span>
                  <span
                    className="exercise-tag"
                  >
                    time:long
                  </span>
                  <span
                    className="exercise-tag"
                  >
                    ID:2@1
                  </span>
                </div>
              </div>
            </div>
          </div>
          <div
            className="openstax exercise-wrapper"
          >
            <div
              className="openstax-exercise-preview exercise-card non-interactive is-vertically-truncated card"
              data-exercise-id="3@1"
              tabIndex={-1}
            >
              <div
                className="card-header"
              >
                <div
                  className="-exercise-header c0"
                >
                  <span
                    className="exercise-number"
                  >
                    3
                  </span>
                </div>
              </div>
              <div
                className="card-body"
              >
                <div
                  className="exercise-body"
                >
                  <div
                    className="openstax-question openstax-question-preview"
                  >
                    <div
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "<b>Meter-area reduction in copper</b> - does it affect the speed of light, etc?",
                        }
                      }
                    />
                    <div
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "In what sense does Einsteinâ€™s theory of relativity illustrate that physics describes fundamental aspects of our universe?",
                        }
                      }
                    />
                    <div
                      className="answers-table"
                    >
                      <div
                        className="openstax-answer"
                      >
                        <section
                          className="answers-answer disabled"
                          role="region"
                        >
                          <div
                            className="correct-incorrect"
                          />
                          <label
                            className="answer-label"
                            htmlFor="3-option-0"
                            onKeyPress={[Function]}
                          >
                            <button
                              aria-label="Choice a:"
                              className="answer-letter"
                              disabled={true}
                            >
                              a
                            </button>
                            <div
                              className="answer-answer"
                            >
                              <div
                                dangerouslySetInnerHTML={
                                  Object {
                                    "__html": "It describes how speed affects different observersâ€™ measurements of time and space.",
                                  }
                                }
                              />
                            </div>
                          </label>
                        </section>
                      </div>
                      <div
                        className="openstax-answer"
                      >
                        <section
                          className="answers-answer disabled"
                          role="region"
                        >
                          <div
                            className="correct-incorrect"
                          />
                          <label
                            className="answer-label"
                            htmlFor="3-option-1"
                            onKeyPress={[Function]}
                          >
                            <button
                              aria-label="Choice b:"
                              className="answer-letter"
                              disabled={true}
                            >
                              b
                            </button>
                            <div
                              className="answer-answer"
                            >
                              <div
                                dangerouslySetInnerHTML={
                                  Object {
                                    "__html": "It describes how a frame of reference is necessary to describe position or motion.",
                                  }
                                }
                              />
                            </div>
                          </label>
                        </section>
                      </div>
                      <div
                        className="openstax-answer"
                      >
                        <section
                          className="answers-answer disabled"
                          role="region"
                        >
                          <div
                            className="correct-incorrect"
                          />
                          <label
                            className="answer-label"
                            htmlFor="3-option-2"
                            onKeyPress={[Function]}
                          >
                            <button
                              aria-label="Choice c:"
                              className="answer-letter"
                              disabled={true}
                            >
                              c
                            </button>
                            <div
                              className="answer-answer"
                            >
                              <div
                                dangerouslySetInnerHTML={
                                  Object {
                                    "__html": "It describes how people think of other peopleâ€™s views from their own frame of reference.",
                                  }
                                }
                              />
                            </div>
                          </label>
                        </section>
                      </div>
                      <div
                        className="openstax-answer"
                      >
                        <section
                          className="answers-answer disabled"
                          role="region"
                        >
                          <div
                            className="correct-incorrect"
                          />
                          <label
                            className="answer-label"
                            htmlFor="3-option-3"
                            onKeyPress={[Function]}
                          >
                            <button
                              aria-label="Choice d:"
                              className="answer-letter"
                              disabled={true}
                            >
                              d
                            </button>
                            <div
                              className="answer-answer"
                            >
                              <div
                                dangerouslySetInnerHTML={
                                  Object {
                                    "__html": "It describes how different parts of the universe are far apart and do not affect each other",
                                  }
                                }
                              />
                            </div>
                          </label>
                        </section>
                      </div>
                      <div
                        className="openstax-answer"
                      >
                        <section
                          className="answers-answer disabled"
                          role="region"
                        >
                          <div
                            className="correct-incorrect"
                          />
                          <label
                            className="answer-label"
                            htmlFor="3-option-4"
                            onKeyPress={[Function]}
                          >
                            <button
                              aria-label="Choice e:"
                              className="answer-letter"
                              disabled={true}
                            >
                              e
                            </button>
                            <div
                              className="answer-answer"
                            >
                              <div
                                dangerouslySetInnerHTML={
                                  Object {
                                    "__html": "It describes how speed affects different observersâ€™ measurements of time and space.",
                                  }
                                }
                              />
                            </div>
                          </label>
                        </section>
                      </div>
                    </div>
                    <div
                      className="detailed-solution"
                    >
                      <div
                        className="header"
                      >
                        Detailed solution
                      </div>
                      <div
                        dangerouslySetInnerHTML={
                          Object {
                            "__html": "four",
                          }
                        }
                      />
                    </div>
                  </div>
                </div>
                <div
                  className="exercise-tags"
                >
                  <span
                    className="exercise-tag"
                  >
                    lo:stax-phys:1-2-1
                  </span>
                  <span
                    className="exercise-tag"
                  >
                    blooms:3
                  </span>
                  <span
                    className="exercise-tag"
                  >
                    dok:3
                  </span>
                  <span
                    className="exercise-tag"
                  >
                    time:long
                  </span>
                  <span
                    className="exercise-tag"
                  >
                    ID:3@1
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="visible-when-debugging"
    >
      <div
        className="tour-spy-info"
      >
        available regions: []; region tour ids: []; audience tags: []; tour tags: []; eligible tours: []; TOUR RIDE: &lt;none&gt;
      </div>
    </div>
  </div>
  <a
    aria-hidden="true"
    className="debug-toggle-link"
    href="#spy"
    onClick={[Function]}
    tabIndex={-1}
  >
    Ï€
  </a>
</div>
`;
